specs/WebMessageObjects.md

```diff
@@ -0,0 +1,355 @@
+ask from WebView2)[https://github.com/MicrosoftEdge/WebView2Feedback/issues/501] is to be able get
+to the paths for DOM file objects, which is not accessible to page content. We also have asks to
+expose other DOM objects, including iframes, <object> objects, etc. which will be added under this
+WebMessageObjects framework in future.
+
+Also in future we want to be able to to inject DOM objects into WebView2 content constructed via the
+app and via the CoreWebView2.PostWebMessage API in the other direction. This API surface needs to be
+compatible with that. (See Appendix)
+
+# Conceptual pages (How To)
+WebMessageObjects are representations of DOM objects that can be passed via the [WebView2 WebMessage
+API](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2webmessagereceivedeventargs).
+You can examine supported DOM objects using native reflections of the types.
+
+Currently the only supported DOM object type with this API is:
+- [File](https://developer.mozilla.org/en-US/docs/Web/API/File)
+
+Page content in WebView2 can pass objects to the app via the
+`chrome.webview.postMessageWithAdditionalObjects(string message, array<object> objects)` content API
+that takes in the array of such supported DOM objects or you can also use
+(ExecuteScript)[https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.executescriptasync]
+with same API. If an invalid or unsupported object is passed via this API, an exception will be
+thrown to the caller and the message will fail to post.
+
+On the WebMessageReceived event handler, the app will retrieve the native representation of objects
+via `AdditionalObjects` property and cast the passed objects to their native types. For example, the
+DOM File object will be exposed as a `CoreWebView2File` object, allowing the app to read the path of
+the passed DOM File object.
+
+# Examples
+## Read the File path of a DOM File
+
+Use this API with a DOM File object to be able to get the path of files dropped on WebView2. The
+HTML and JavaScript snippets are part of the page content code running inside WebView2 and are used
+by both the C++ and C# sample code below.
+
+```html
+<!-- File upload location -->
+<input type="file" id="files" multiple />
+```
+
+```javascript
+const input = document.getElementById('files');
+input.addEventListener('change', function() {
+    const currentFiles = input.files;
+    chrome.webview.postMessageWithAdditionalObjects("FilesDropped", currentFiles);
+});
+```
+
+```cpp
+m_webView->add_WebMessageReceived(
+    Callback<ICoreWebView2WebMessageReceivedEventHandler>(
+        [this](ICoreWebView2* sender, ICoreWebView2WebMessageReceivedEventArgs* args)
+        {
+            wil::com_ptr<ICoreWebView2WebMessageReceivedEventArgs2> args2 =
+                wil::com_ptr<ICoreWebView2WebMessageReceivedEventArgs>(args)
+                    .query<ICoreWebView2WebMessageReceivedEventArgs2>();

```

Comment:
I vaguely recall that the policy is that C++ samples are version-agile, whereas the C# samples are version-locked. But here, we are being version-locked in the C++ sample.

Reply:
Yes, please fix to allow for older WebView2 Runtimes that haven't implemented the new event args.